controllers/state_of_the_world.go

```diff
@@ -64,14 +72,137 @@ func NewPolicyMachineryController(manager ctrlruntime.Manager, client *dynamic.D
 
 func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
 	reconciler := &controller.Workflow{
-		Precondition: NewEventLogger().Log,
+		Precondition: (&controller.Workflow{
+			Precondition: NewEventLogger().Log,
+			Tasks: []controller.ReconcileFunc{
+				NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			},
+		}).Run,
 		Tasks: []controller.ReconcileFunc{
-			NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			NewAuthorinoCrReconciler(client).Subscription().Reconcile,
 		},
 	}
 	return reconciler.Run
 }
 
+type AuthorinoCrReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewAuthorinoCrReconciler(client *dynamic.DynamicClient) *AuthorinoCrReconciler {
+	return &AuthorinoCrReconciler{Client: client}
+}
+
+func (r *AuthorinoCrReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{
+		ReconcileFunc: r.Reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: ptr.To(kuadrantv1beta1.KuadrantKind), EventType: ptr.To(controller.CreateEvent)},
+			{Kind: ptr.To(kuadrantv1beta1.AuthorinoKind), EventType: ptr.To(controller.DeleteEvent)},
+		},
+	}
+}
+
+func (r *AuthorinoCrReconciler) Reconcile(ctx context.Context, events []controller.ResourceEvent, topology *machinery.Topology, _ error) {
+	logger := controller.LoggerFromContext(ctx).WithName("AuthorinoCrReconciler")
+	logger.Info("Reconciling Authorino Cr")
+
+	kobj := &kuadrantv1beta1.Kuadrant{}
+	for _, event := range events {
+		if event.Kind == kuadrantv1beta1.KuadrantKind && event.EventType == controller.CreateEvent {
```

Comment:
Sorry @Boomatang. I'm not convinced about this approach of matching the event type again, to decide how to search for the Kuadrant object in the topology.

IMO, the first root that matches the Kuadrant kind is the Kuadrant object we are looking for. If there are more than one, then only the first Kuadrant should get an Authorino instance created for it. If there is none, than no Authorino object should be created and all existing Authorino instances should be deleted.

Reply:
Ya looking at it again you are right there was no need to act depending on the event kind/type. That will only matter when we allow multiple Kuadrant resources. Refactored to be some what cleaner I think.